### PR TITLE
Add date range and format date operators

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?
- Added `date-range` and `format-date` operator
- Small refactor to `Operator`/`WhenRuleEvaluator` so that operator is responsible for the operations.
- `WhenRuleEvaluator` will be renamed to `RuleEvaluator` in the next PR, as it is not specific to when rules.

### How to review 
Ensure the changes are appropriate.

Depends on: https://github.com/ONSdigital/eq-questionnaire-runner/pull/725

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

File: app/questionnaire/routing/operator.py [hunk 1]
```diff
@@ -50,3 +53,23 @@ def evaluate(
     @staticmethod
     def _any_operands_none(*operands: Union[Sequence, ValueTypes]) -> bool:
         return any(operand is None for operand in operands)
+
+
+OPERATION_MAPPING: dict[str, str] = {
```
Comment (LJBabbage) [inline app/questionnaire/routing/operator.py:58]: I can live with it, but I as you probably had guess I wasn't a huge fan of mapping the operation to the function name with a string, it felt to me more likely to break, but did add more padding.

**Review by LJBabbage — APPROVED** [review]

File: app/questionnaire/routing/operator.py (hunk 1 above)
Comment (MebinAbraham) [reply app/questionnaire/routing/operator.py:58]: Understandable, I mainly changed it so rules evaluator is not responsible for the operation. The str mapping is a consequence of our operator names, as we can't have a 1:1 method name like placeholder transforms. However, we could still do what placeholder does, but just make a mapping for the operators that are "special" i.e `!=` etc. Don't think it's worth changes the operator name itself for the sake of python.

File: app/questionnaire/routing/operations.py [hunk 2]
```diff
@@ -33,6 +35,10 @@ class Operations:
     A class to group the operations
     """
 
+    def __init__(self, language: str):
+        self._language = language
+        self._locale = DEFAULT_LOCALE if language in ["en", "eo"] else language
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/questionnaire/routing/operations.py:40]: We don't enforce type on `__int__` but have added anyway.

File: app/questionnaire/routing/operator.py [hunk 3]
```diff
@@ -19,10 +20,12 @@ class Operator:
...
     COUNT: str = "count"
     DATE: str = "date"
+    DATE_RANGE: str = "date-range"
+    FORMAT_DATE: str = "format-date"
 
-    def __init__(self, name: str, operation: Callable) -> None:
+    def __init__(self, name: str, operations: Operations) -> None:
         self.name = name
-        self._operation = operation
+        self._operation = getattr(operations, OPERATION_MAPPING[self.name])
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/questionnaire/routing/operator.py:28]: It is no longer a callable as the operation is not instantiated at module level. They could be callable if the mapping was in rule evaluator class, but operations shouldn't be coupled with that. The only reason we need the mapping is cause we don't have a 1:1 with operator to func names as some of our operator wouldn't be supported i.e `==` etc, but we shouldn't change the operator for the sake of some python.

File: app/questionnaire/routing/operator.py [hunk 4]
```diff
@@ -50,3 +53,23 @@ def evaluate(
     @staticmethod
     def _any_operands_none(*operands: Union[Sequence, ValueTypes]) -> bool:
         return any(operand is None for operand in operands)
+
+
+OPERATION_MAPPING: dict[str, str] = {
+    Operator.NOT: "evaluate_not",
+    Operator.AND: "evaluate_and",
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply app/questionnaire/routing/operator.py:60]: https://github.com/ONSdigital/eq-questionnaire-runner/pull/726#discussion_r758167368

**Review by parimooa — APPROVED** [review]

**Review by LJBabbage — APPROVED** [review]

**Pull request merged**